### PR TITLE
Fix app update and duplicate removal

### LINKED ISSUE #2884: After Nextcloud or polls update, activating polls failed (Integrity constraint violation: 1062 Duplicate entry)
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### What went wrong, what did you observe?

I upgrade Nextcloud from 24.0.12 to 25.0.6 on a Debian 11 with MySQL, everything is OK. But polls (5.0.4) was disabled, if I enable it, this error occure:

```
[settings] Fehler: Exception: Database error when running migration 040102Date20230123072601 for app polls
An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '2--0' for key 'UNIQ_options' at <<closure>>

0. /var/www/html/nextcloud/lib/private/Installer.php line 153
   OC\DB\MigrationService->migrate()
1. /var/www/html/nextcloud/apps/settings/lib/Controller/AppSettingsController.php line 451
   OC\Installer->installApp()
2. /var/www/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 225
   OCA\Settings\Controller\AppSettingsController->enableApps()
3. /var/www/html/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 133
   OC\AppFramework\Http\Dispatcher->executeController()
4. /var/www/html/nextcloud/lib/private/AppFramework/App.php line 172
   OC\AppFramework\Http\Dispatcher->dispatch()
5. /var/www/html/nextcloud/lib/private/Route/Router.php line 298
   OC\AppFramework\App::main()
6. /var/www/html/nextcloud/lib/base.php line 1048
   OC\Route\Router->match()
7. /var/www/html/nextcloud/index.php line 36
   OC::handleRequest()

POST /cloud/settings/apps/enable
from 80.xxx.xxx.xxx by BurtGummer at 2023-04-27T12:03:25+02:00
```

I try to disable/enable it, it seems its the same problem from this issue: https://github.com/nextcloud/polls/issues/2853
but there is no change.

### What did you expect, how polls should behave instead?

Just enable the app polls:

root@server /var/www/html/nextcloud # sudo -u www-data php occ app:update polls
root@server /var/www/html/nextcloud # cd /var/www/html/nextcloud/ ; sudo -u www-data php occ app:enable polls
Database error when running migration 040102Date20230123072601 for app polls
An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '2--0' for key 'UNIQ_options'
root@server /var/www/html/nextcloud #

### What steps does it need to replay this bug?

1. before the update, polls works great.

### Installation method

Installed/updated from the appstore (Apps section of your site)

### Installation type

Updated from a minor version within same major version (i.e. 4.0.0 to 4.1.1)

### Affected polls version

5.0.4

### Which browser did you use, when experiencing the bug?

- [X] Firefox
- [ ] Chrome
- [ ] Chromium/Chromium based (i.e. Edge)
- [ ] Safari
- [ ] Other/Don't know

### Which System did you use, when experiencing the bug?

- [ ] Windows
- [ ] Linux
- [ ] MacOs
- [ ] iOS
- [ ] Android
- [ ] Other/Don't know

Nextcloud 25 (Nextcloud Hub 3)

### PHP engine version

PHP 7.3

### Database engine

MySQL

### Which user-backends are you using?

- [ ] Default user-backend _(database)_
- [X] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other/Don't know

### Configuration report

```shell
root@server /var/www/html/nextcloud # sudo -u www-data php occ config:list system
{
    "system": {
        "updatechecker": false,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "oc.example.org"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/oc.example.org\/cloud\/",
        "dbtype": "mysql",
        "version": "25.0.6.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "Europe\/Berlin",
        "installed": true,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "sendmail",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "theme": "",
        "loglevel": 3,
        "lost_password_link": "disabled",
        "default_phone_region": "DE",
        "maintenance": false,
        "ldapIgnoreNamingRules": false,
        "ldapUserCleanupInterval": 60,
        "htaccess.RewriteBase": "\/cloud",
        "updater.release.channel": "stable",
        "appstore.experimental.enabled": false,
        "overwriteprotocol": "https",
        "forcessl": true,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "mail_sendmailmode": "smtp",
        "mysql.utf8mb4": true
    }
}

root@server /var/www/html/nextcloud # sudo -u www-data php occ app:list
Enabled:
  - activity: 2.17.0
  - admin_audit: 1.15.0
  - circles: 25.0.0
  - cloud_federation_api: 1.8.0
  - comments: 1.15.0
  - contactsinteraction: 1.6.0
  - dav: 1.24.0
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_pdfviewer: 2.6.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - firstrunwizard: 2.14.0
  - impersonate: 1.12.0
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - nextcloud_announcements: 1.14.0
  - notifications: 2.13.1
  - oauth2: 1.13.0
  - password_policy: 1.15.0
  - photos: 2.0.1
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - recommendations: 1.4.0
  - related_resources: 1.0.4
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - support: 1.8.0
  - survey_client: 1.13.0
  - text: 3.6.0
  - theming: 2.0.1
  - twofactor_backupcodes: 1.14.0
  - updatenotification: 1.15.0
  - user_ldap: 1.15.0
  - viewer: 1.9.0
  - workflowengine: 2.7.0
Disabled:
  - bruteforcesettings: 1.1.0
  - dashboard: 7.0.0
  - encryption
  - files_external
  - files_rightclick: 0.8.4
  - keeweb: 0.6.12
  - polls: 3.8.4
  - suspicious_login
  - systemtags: 1.15.0
  - twofactor_totp
  - user_status: 1.0.1
  - weather_status: 1.2.0
root@server /var/www/html/nextcloud
```

### PR DESCRIPTION
**Pull request by dartcafe**
fixes #2884
* move duplicate removal to DBAL
* make sure, duplicates get removed before creating indexes

### First time install 
Nextcloud executes
* unexecuteted `migration classes` (not listed in the `*_migrations` table) and the 
* `install` repair step.

### Version update 
Nextcloud executes 
* `pre-migration` repair steps, 
* unexecuteted `migration classes` (not listed in the `*_migrations` table) and the 
* `post-migration` repair steps

### App enabling after just disabling the already installed app 
Nextcloud executes 
* `pre-migration` repair steps, 
* unexecuteted `migration classes` (not listed in the `*_migrations` table) and the
* `post-migration` repair steps and the 
* `install` repair step

❗ As a compromise at the moment we allow the index creation to be ran twice when enabling the app via app store or `occ`, to ensure all indexes are created properly for every install/update/enabling path.

## Removing Polls from instance
A command `occ polls:db:purge` to remove Polls completely was added in this PR. 
* remove all Polls' tables
* remove all Polls' migration records
* remove all Polls' app config records (this also disables Polls)

This does not remove Polls' files (call `occ app:remove polls` to remove it complete afterwards) but it resets Polls into an 'uninstalled' state. Enabling the app is then equivalent to a first time install and calls the migration and the install repair step (see above).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -69,9 +70,41 @@ If you want to run the latest development version from git source, you need to c
...
 ```
 
-* Install dev environment with ```make setup-dev```
+* Install dev environment with ```make setup-dev``` or
+* install runtime environment with ```make setup-build```
 * Compile javascript with ```npm run build```
 * Run a complete build with ```make appstore``` (Find the output in the build directory)
+* call `occ app:enable polls` to enable Polls
+
+### Installation variants
```
Comment (bencomp) [inline README.md:79]: The header level is the same as the next subsections' header level; should they be different?

File: lib/Migration/RepairSteps/Install.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * @copyright Copyright (c) 2021 René Gieling <github@dartcafe.de>
```
Comment (bencomp) [inline lib/Migration/RepairSteps/Install.php:3]: Should the year be 2023 here?

File: README.md [hunk 3]
```diff
@@ -69,9 +70,41 @@ If you want to run the latest development version from git source, you need to c
...
+
+### First time install
+Nextcloud executes
+* unexecuteted `migration classes` (not listed in the `*_migrations` table) and the
```
Comment (bencomp) [inline README.md:83]: "unexecuteted" should be "unexecuted".

**Review by bencomp — COMMENTED** [review]
Thanks for your work! I have a few minor after-the-merge comments.

**Comment by dartcafe** [comment]
Thanks for watching that. If you mind, you may correct that in a new PR. I am tired for now, spent more time than planned with fiddeling out the different app activating paths. 😉

**Pull request merged**